# Maji Ndogo Water Project — Data Cleaning and Preparation

## Mission Context

> **From:** Aziza Naledi, Project Lead  
> **To:** Data Analysis Team  
>
> Our mission, as arduous as it is essential, requires us to delve deeper into our reservoir of data.  
> To truly illuminate the road ahead, we must magnify our analysis — moving beyond isolated data points to discern larger patterns and trends.  
>  
> In this next step, we will cluster our data to gain a panoramic understanding.  
> Every insight we uncover brings us one stride closer to solving **Maji Ndogo's water crisis**.  



## Getting Started

Before diving into transformations, we’ll scan through the **data dictionary** and **sample key tables** to get a feel for the structure of our database.

This step helps us confirm column names, relationships, and identify any inconsistencies that might affect our later analysis.


In [10]:
%%sql

SELECT *
FROM data_dictionary
LIMIT 10;


 * mysql+pymysql://root:***@localhost/md_water_services
10 rows affected.


[('employee', 'assigned_employee_id', 'Unique ID assigned to each employee', 'INT', 'visits'),
 ('employee', 'employee_name', 'Name of the employee', 'VARCHAR(255)', ''),
 ('employee', 'phone_number', 'Contact number of the employee', 'VARCHAR(15)', ''),
 ('employee', 'email', 'Email address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'address', 'Residential address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'town_name', 'Name of the town where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'province_name', 'Name of the province where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'position', 'Position or job title of the employee', 'VARCHAR(255)', ''),
 ('visits', 'record_id', 'Unique ID assigned to each visit', 'int', 'water_quality, water_source'),
 ('visits', 'location_id', 'ID of the location visited', 'varchar(255)', 'location')]

### Exploring the `employee` Table

Our next focus is on the **employee** table.  
It contains information about all staff members working on the Maji Ndogo project.  

However, the **email addresses are missing**, and we’ll need them to send out reports and updates.  
Fortunately, every staff member’s email follows a simple pattern:

`first_name.last_name@ndogowater.gov`

Before we update the table, we’ll generate this format using SQL functions to ensure it looks correct.


In [11]:
%%sql

SELECT *
FROM employee
LIMIT 5;


 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287 ', None, '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786 ', None, '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041 ', None, '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900 ', None, '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640 ', None, '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

### Step 1 — Replace Spaces with Dots

We begin by transforming the `employee_name` column to replace spaces with a period (`.`) between first and last names.  
We’ll use the **REPLACE()** function to achieve this.


In [12]:
%%sql

SELECT
    REPLACE(employee_name, ' ', '.') AS name_with_dot
FROM employee;


 * mysql+pymysql://root:***@localhost/md_water_services
56 rows affected.


[('Amara.Jengo',),
 ('Bello.Azibo',),
 ('Bakari.Iniko',),
 ('Malachi.Mavuso',),
 ('Cheche.Buhle',),
 ('Zuriel.Matembo',),
 ('Deka.Osumare',),
 ('Lalitha.Kaburi',),
 ('Enitan.Zuri',),
 ('Farai.Nia',),
 ('Gamba.Shani',),
 ('Harith.Nyota',),
 ('Isoke.Amani',),
 ('Jengo.Tumaini',),
 ('Kunto.Asha',),
 ('Lesedi.Kofi',),
 ('Makena.Thabo',),
 ('Nia.Furaha',),
 ('Ona.Sefu',),
 ('Pili.Zola',),
 ('Quibilah.Akua',),
 ('Rudo.Imani',),
 ('Sanaa.Tendaji',),
 ('Thandiwe.Kito',),
 ('Usafi.Ayo',),
 ('Vuyisile.Kwame',),
 ('Wambui.Jabali',),
 ('Xola.Uzuri',),
 ('Yewande.Ebele',),
 ('Nadia.Zalika',),
 ('Ayo.Deka',),
 ('Bahira.Jamila',),
 ('Chidi.Kunto',),
 ('Dalila.Lesedi',),
 ('Eshe.Makena',),
 ('Faridah.Khalida',),
 ('Gavivi.Ona',),
 ('Hasani.Pili',),
 ('Iniko.Quibilah',),
 ('Jengo.Rudo',),
 ('Kofi.Sanaa',),
 ('Lulu.Thandiwe',),
 ('Makena.Usafi',),
 ('Nia.Vuyisile',),
 ('Ona.Wambui',),
 ('Pili.Xola',),
 ('Quibilah.Yewande',),
 ('Rudo.Zalika',),
 ('Sanaa.Ayo',),
 ('Thabo.Akida',),
 ('Usafi.Bahir',),
 ('Vuyisile.Ghadir',),
 ('Wambui.Hadiya',),
 ('Xola.Inaya',),
 ('Yewande.Jengo',),
 ('Zuri.Kofi',)]

### Step 2 — Convert to Lowercase

Next, we make all characters lowercase using the **LOWER()** function.  
This ensures our email format stays consistent across all entries.


In [13]:
%%sql

SELECT
    LOWER(REPLACE(employee_name, ' ', '.')) AS lowercase_name
FROM employee;


 * mysql+pymysql://root:***@localhost/md_water_services
56 rows affected.


[('amara.jengo',),
 ('bello.azibo',),
 ('bakari.iniko',),
 ('malachi.mavuso',),
 ('cheche.buhle',),
 ('zuriel.matembo',),
 ('deka.osumare',),
 ('lalitha.kaburi',),
 ('enitan.zuri',),
 ('farai.nia',),
 ('gamba.shani',),
 ('harith.nyota',),
 ('isoke.amani',),
 ('jengo.tumaini',),
 ('kunto.asha',),
 ('lesedi.kofi',),
 ('makena.thabo',),
 ('nia.furaha',),
 ('ona.sefu',),
 ('pili.zola',),
 ('quibilah.akua',),
 ('rudo.imani',),
 ('sanaa.tendaji',),
 ('thandiwe.kito',),
 ('usafi.ayo',),
 ('vuyisile.kwame',),
 ('wambui.jabali',),
 ('xola.uzuri',),
 ('yewande.ebele',),
 ('nadia.zalika',),
 ('ayo.deka',),
 ('bahira.jamila',),
 ('chidi.kunto',),
 ('dalila.lesedi',),
 ('eshe.makena',),
 ('faridah.khalida',),
 ('gavivi.ona',),
 ('hasani.pili',),
 ('iniko.quibilah',),
 ('jengo.rudo',),
 ('kofi.sanaa',),
 ('lulu.thandiwe',),
 ('makena.usafi',),
 ('nia.vuyisile',),
 ('ona.wambui',),
 ('pili.xola',),
 ('quibilah.yewande',),
 ('rudo.zalika',),
 ('sanaa.ayo',),
 ('thabo.akida',),
 ('usafi.bahir',),
 ('vuyisile.ghadir',),
 ('wambui.hadiya',),
 ('xola.inaya',),
 ('yewande.jengo',),
 ('zuri.kofi',)]

### Step 3 — Concatenate the Email Domain

Finally, we use the **CONCAT()** function to add the organization’s domain:  
`@ndogowater.gov`  
This produces the final email address format we’ll later update in the database.


In [14]:
%%sql

SELECT
    CONCAT(
        LOWER(REPLACE(employee_name, ' ', '.')),
        '@ndogowater.gov'
    ) AS new_email
FROM employee;


 * mysql+pymysql://root:***@localhost/md_water_services
56 rows affected.


[('amara.jengo@ndogowater.gov',),
 ('bello.azibo@ndogowater.gov',),
 ('bakari.iniko@ndogowater.gov',),
 ('malachi.mavuso@ndogowater.gov',),
 ('cheche.buhle@ndogowater.gov',),
 ('zuriel.matembo@ndogowater.gov',),
 ('deka.osumare@ndogowater.gov',),
 ('lalitha.kaburi@ndogowater.gov',),
 ('enitan.zuri@ndogowater.gov',),
 ('farai.nia@ndogowater.gov',),
 ('gamba.shani@ndogowater.gov',),
 ('harith.nyota@ndogowater.gov',),
 ('isoke.amani@ndogowater.gov',),
 ('jengo.tumaini@ndogowater.gov',),
 ('kunto.asha@ndogowater.gov',),
 ('lesedi.kofi@ndogowater.gov',),
 ('makena.thabo@ndogowater.gov',),
 ('nia.furaha@ndogowater.gov',),
 ('ona.sefu@ndogowater.gov',),
 ('pili.zola@ndogowater.gov',),
 ('quibilah.akua@ndogowater.gov',),
 ('rudo.imani@ndogowater.gov',),
 ('sanaa.tendaji@ndogowater.gov',),
 ('thandiwe.kito@ndogowater.gov',),
 ('usafi.ayo@ndogowater.gov',),
 ('vuyisile.kwame@ndogowater.gov',),
 ('wambui.jabali@ndogowater.gov',),
 ('xola.uzuri@ndogowater.gov',),
 ('yewande.ebele@ndogowater.gov',),
 ('nadia.zalika@ndogowater.gov',),
 ('ayo.deka@ndogowater.gov',),
 ('bahira.jamila@ndogowater.gov',),
 ('chidi.kunto@ndogowater.gov',),
 ('dalila.lesedi@ndogowater.gov',),
 ('eshe.makena@ndogowater.gov',),
 ('faridah.khalida@ndogowater.gov',),
 ('gavivi.ona@ndogowater.gov',),
 ('hasani.pili@ndogowater.gov',),
 ('iniko.quibilah@ndogowater.gov',),
 ('jengo.rudo@ndogowater.gov',),
 ('kofi.sanaa@ndogowater.gov',),
 ('lulu.thandiwe@ndogowater.gov',),
 ('makena.usafi@ndogowater.gov',),
 ('nia.vuyisile@ndogowater.gov',),
 ('ona.wambui@ndogowater.gov',),
 ('pili.xola@ndogowater.gov',),
 ('quibilah.yewande@ndogowater.gov',),
 ('rudo.zalika@ndogowater.gov',),
 ('sanaa.ayo@ndogowater.gov',),
 ('thabo.akida@ndogowater.gov',),
 ('usafi.bahir@ndogowater.gov',),
 ('vuyisile.ghadir@ndogowater.gov',),
 ('wambui.hadiya@ndogowater.gov',),
 ('xola.inaya@ndogowater.gov',),
 ('yewande.jengo@ndogowater.gov',),
 ('zuri.kofi@ndogowater.gov',)]

## Updating Employee Email Addresses

Now that we have verified the email format, it’s time to update the `employee` table and store the generated email addresses.

We’ll use an **UPDATE** query that applies the same transformation logic we tested earlier:

- Replace spaces in `employee_name` with dots (`.`)
- Convert all letters to lowercase
- Append the official domain: `@ndogowater.gov`

This ensures that every employee record now has a valid and consistent email address.


In [15]:
%%sql

UPDATE employee
SET email = CONCAT(
    LOWER(REPLACE(employee_name, ' ', '.')),
    '@ndogowater.gov'
);


 * mysql+pymysql://root:***@localhost/md_water_services
56 rows affected.


[]

### Verifying the Email Update

After updating the table, we should confirm that the `email` column has been populated correctly.  
We’ll select the first few records to visually verify the change.


In [16]:
%%sql

SELECT employee_name, email
FROM employee
LIMIT 10;


 * mysql+pymysql://root:***@localhost/md_water_services
10 rows affected.


[('Amara Jengo', 'amara.jengo@ndogowater.gov'),
 ('Bello Azibo', 'bello.azibo@ndogowater.gov'),
 ('Bakari Iniko', 'bakari.iniko@ndogowater.gov'),
 ('Malachi Mavuso', 'malachi.mavuso@ndogowater.gov'),
 ('Cheche Buhle', 'cheche.buhle@ndogowater.gov'),
 ('Zuriel Matembo', 'zuriel.matembo@ndogowater.gov'),
 ('Deka Osumare', 'deka.osumare@ndogowater.gov'),
 ('Lalitha Kaburi', 'lalitha.kaburi@ndogowater.gov'),
 ('Enitan Zuri', 'enitan.zuri@ndogowater.gov'),
 ('Farai Nia', 'farai.nia@ndogowater.gov')]

## Cleaning Phone Numbers

During data collection and migration, minor inconsistencies often appear — such as extra spaces or incorrect lengths in phone numbers.

To confirm this, we’ll inspect the number of characters in the `phone_number` column using the **LENGTH()** function.

Expected format:  +99XXXXXXXXX

That’s **12 characters total** — but our check will show **13**, meaning there’s an extra space at the end.



In [17]:
%%sql

SELECT
    phone_number,
    LENGTH(phone_number) AS phone_length
FROM employee
LIMIT 10;


 * mysql+pymysql://root:***@localhost/md_water_services
10 rows affected.


[('+99637993287 ', 13),
 ('+99643864786 ', 13),
 ('+99222599041 ', 13),
 ('+99945849900 ', 13),
 ('+99381679640 ', 13),
 ('+99034075111 ', 13),
 ('+99379364631 ', 13),
 ('+99681623240 ', 13),
 ('+99248509202 ', 13),
 ('+99570082739 ', 13)]

### Trimming Extra Spaces

The issue arises from an unwanted space at the end of each phone number.  
To fix this, we’ll use the **TRIM()** function, which removes any leading or trailing spaces in a string.

Let’s first preview the result to ensure the trimming works correctly before applying it to the full dataset.


In [18]:
%%sql

SELECT
    phone_number AS original,
    TRIM(phone_number) AS trimmed
FROM employee
LIMIT 10;


 * mysql+pymysql://root:***@localhost/md_water_services
10 rows affected.


[('+99637993287 ', '+99637993287'),
 ('+99643864786 ', '+99643864786'),
 ('+99222599041 ', '+99222599041'),
 ('+99945849900 ', '+99945849900'),
 ('+99381679640 ', '+99381679640'),
 ('+99034075111 ', '+99034075111'),
 ('+99379364631 ', '+99379364631'),
 ('+99681623240 ', '+99681623240'),
 ('+99248509202 ', '+99248509202'),
 ('+99570082739 ', '+99570082739')]

### Applying the Fix

Now that we’ve verified `TRIM()` works, we can update the `employee` table to permanently clean all phone numbers.


In [19]:
%%sql

UPDATE employee
SET phone_number = TRIM(phone_number);


 * mysql+pymysql://root:***@localhost/md_water_services
56 rows affected.


[]

### Clean phone numbers and count employees by town

Steps:
1. Inspect the `employee` table structure to confirm the column names (so we can find the town/city column).  
2. Preview the effect of `TRIM()` on `phone_number` to ensure trailing spaces are removed.  
3. Update the `phone_number` column to permanently remove leading/trailing spaces.  
4. Verify the update worked.  
5. Count how many employees live in each town (we’ll use `COALESCE()` to handle common possible column names for town/city).


In [20]:
%%sql

DESCRIBE employee;


 * mysql+pymysql://root:***@localhost/md_water_services
8 rows affected.


[('assigned_employee_id', 'int', 'NO', 'PRI', None, ''),
 ('employee_name', 'varchar(255)', 'YES', '', None, ''),
 ('phone_number', 'varchar(15)', 'YES', '', None, ''),
 ('email', 'varchar(255)', 'YES', '', None, ''),
 ('address', 'varchar(255)', 'YES', '', None, ''),
 ('province_name', 'varchar(255)', 'YES', '', None, ''),
 ('town_name', 'varchar(255)', 'YES', '', None, ''),
 ('position', 'varchar(255)', 'YES', '', None, '')]

#### Preview TRIM() on phone_number

We preview the trimmed phone number next to the original so we can confirm trimming is correct before updating the table.


In [23]:
%%sql

SELECT
    assigned_employee_id,
    employee_name,
    phone_number AS original_phone,
    TRIM(phone_number) AS trimmed_phone,
    LENGTH(phone_number) AS original_len,
    LENGTH(TRIM(phone_number)) AS trimmed_len
FROM employee
LIMIT 10;


 * mysql+pymysql://root:***@localhost/md_water_services
10 rows affected.


[(0, 'Amara Jengo', '+99637993287', '+99637993287', 12, 12),
 (1, 'Bello Azibo', '+99643864786', '+99643864786', 12, 12),
 (2, 'Bakari Iniko', '+99222599041', '+99222599041', 12, 12),
 (3, 'Malachi Mavuso', '+99945849900', '+99945849900', 12, 12),
 (4, 'Cheche Buhle', '+99381679640', '+99381679640', 12, 12),
 (5, 'Zuriel Matembo', '+99034075111', '+99034075111', 12, 12),
 (6, 'Deka Osumare', '+99379364631', '+99379364631', 12, 12),
 (7, 'Lalitha Kaburi', '+99681623240', '+99681623240', 12, 12),
 (8, 'Enitan Zuri', '+99248509202', '+99248509202', 12, 12),
 (10, 'Farai Nia', '+99570082739', '+99570082739', 12, 12)]

#### Update phone_number to remove leading/trailing spaces

If the preview looks good, we update the `phone_number` column using `TRIM()` so all values are cleaned in place.


In [24]:
%%sql

UPDATE employee
SET phone_number = TRIM(phone_number)
WHERE phone_number IS NOT NULL;


 * mysql+pymysql://root:***@localhost/md_water_services
56 rows affected.


[]

#### Verify phone number cleanup

Confirm the lengths are now as expected (e.g., 12 characters for the `+99...` format) and show a few cleaned values.


In [26]:
%%sql

SELECT
    assigned_employee_id,
    employee_name,
    phone_number,
    LENGTH(phone_number) AS length_after_trim
FROM employee
LIMIT 10;


 * mysql+pymysql://root:***@localhost/md_water_services
10 rows affected.


[(0, 'Amara Jengo', '+99637993287', 12),
 (1, 'Bello Azibo', '+99643864786', 12),
 (2, 'Bakari Iniko', '+99222599041', 12),
 (3, 'Malachi Mavuso', '+99945849900', 12),
 (4, 'Cheche Buhle', '+99381679640', 12),
 (5, 'Zuriel Matembo', '+99034075111', 12),
 (6, 'Deka Osumare', '+99379364631', 12),
 (7, 'Lalitha Kaburi', '+99681623240', 12),
 (8, 'Enitan Zuri', '+99248509202', 12),
 (10, 'Farai Nia', '+99570082739', 12)]

### Count employees by town

We’ll count how many employees live in each town. Different datasets call this column different names (e.g., `town`, `city`, `location`, `residence`, `address`), so we use `COALESCE()` to pick the first non-null value among common candidates. We also normalize the result (trim + lower) to avoid duplicates due to casing or stray spaces.


In [34]:
%%sql

SELECT
    COALESCE(
        NULLIF(TRIM(LOWER(town_name)), ''),
        NULLIF(TRIM(LOWER(province_name)), ''),
        NULLIF(TRIM(LOWER(address)), ''),
        'UNKNOWN'
    ) AS location,
    COUNT(*) AS employee_count
FROM employee
GROUP BY location
ORDER BY employee_count DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
9 rows affected.


[('rural', 29),
 ('dahabu', 6),
 ('harare', 5),
 ('lusaka', 4),
 ('zanzibar', 4),
 ('ilanga', 3),
 ('serowe', 3),
 ('kintampo', 1),
 ('yaounde', 1)]

### Finding the Top 3 Field Surveyors

To recognize the top-performing field surveyors, we’ll analyze the `visits` table.  
Each record in this table represents a location visit made by an employee.  
By counting how many visits each employee completed, we can identify who contributed the most to our survey work.

**Steps:**
1. Use the `visits` table to count the total number of visits per employee.  
2. Sort the results from highest to lowest.  
3. Limit the output to show only the top 3 employees.


In [4]:
%%sql

SELECT
    assigned_employee_id,
    COUNT(location_id) AS total_visits
FROM visits
GROUP BY assigned_employee_id
ORDER BY total_visits DESC
LIMIT 3;

 * mysql+pymysql://root:***@localhost/md_water_services
3 rows affected.


[(1, 3708), (30, 3676), (34, 3539)]

### Retrieving the Top 3 Field Surveyors’ Details

Now that we’ve identified the top 3 surveyors by the number of visits, we’ll use their `employee_id` values to look up their personal information.

We’ll extract their:
- Full names  
- Email addresses  
- Phone numbers  

This will allow us to send recognition messages to those who contributed the most to the Maji Ndogo survey.


In [7]:
%%sql

SELECT
    assigned_employee_id,
    employee_name,
    email,
    phone_number
FROM employee
WHERE assigned_employee_id IN (1, 30, 34);


 * mysql+pymysql://root:***@localhost/md_water_services
3 rows affected.


[(1, 'Bello Azibo', 'bello.azibo@ndogowater.gov', '+99643864786'),
 (30, 'Pili Zola', 'pili.zola@ndogowater.gov', '+99822478933'),
 (34, 'Rudo Imani', 'rudo.imani@ndogowater.gov', '+99046972648')]

### Exploring Water Source Distribution Across Maji Ndogo

With our team recognized, let’s now turn our attention to the main objective — understanding where Maji Ndogo’s water sources are located.

We’ll use the `location` table, focusing on:
- `province_name` — the larger administrative region  
- `town_name` — the town within each province  
- `location_type` — whether the area is urban, rural, etc.  

Our goal in this step is to count how many water source records are available per town.  
This will give us an idea of where the water sources are concentrated.


In [8]:
%%sql

SELECT
    town_name,
    COUNT(*) AS num_records
FROM location
GROUP BY town_name
ORDER BY num_records DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
25 rows affected.


[('Rural', 23740),
 ('Harare', 1650),
 ('Amina', 1090),
 ('Lusaka', 1070),
 ('Mrembo', 990),
 ('Asmara', 930),
 ('Dahabu', 930),
 ('Kintampo', 780),
 ('Ilanga', 780),
 ('Isiqalo', 770),
 ('Zuri', 750),
 ('Amara', 710),
 ('Majengo', 620),
 ('Pwani', 520),
 ('Kofi', 510),
 ('Serowe', 480),
 ('Bahari', 470),
 ('Marang', 460),
 ('Abidjan', 400),
 ('Bello', 400),
 ('Cheche', 370),
 ('Deka', 350),
 ('Djenne', 310),
 ('Zanzibar', 300),
 ('Yaounde', 270)]

### Counting Water Source Records per Province

To gain a higher-level view of water source distribution, we’ll summarize the data by **province** instead of town.

This helps us:
- Understand which provinces have the most (or fewest) recorded water sources.
- Identify potential imbalances or areas needing further investigation.


In [9]:
%%sql

SELECT
    province_name,
    COUNT(*) AS num_records
FROM location
GROUP BY province_name
ORDER BY num_records DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[('Kilimani', 9510),
 ('Akatsi', 8940),
 ('Sokoto', 8220),
 ('Amanzi', 6950),
 ('Hawassa', 6030)]

### Counting Records per Town within Each Province

Now that we’ve examined province-level and town-level distributions separately, we’ll combine them into a single, structured summary.

**Goal:**
- Display how many records each **town** has within its **province**.
- Group data by both `province_name` and `town_name`.
- Sort by `province_name` alphabetically, and within each province, order towns by the number of records in **descending** order.

This will help us visualize how water sources are distributed within provinces and identify which towns have the highest concentration of sources.


In [10]:
%%sql

SELECT
    province_name,
    town_name,
    COUNT(*) AS records_per_town
FROM location
GROUP BY province_name, town_name
ORDER BY province_name ASC, records_per_town DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
31 rows affected.


[('Akatsi', 'Rural', 6290),
 ('Akatsi', 'Lusaka', 1070),
 ('Akatsi', 'Harare', 800),
 ('Akatsi', 'Kintampo', 780),
 ('Amanzi', 'Rural', 3100),
 ('Amanzi', 'Asmara', 930),
 ('Amanzi', 'Dahabu', 930),
 ('Amanzi', 'Amina', 670),
 ('Amanzi', 'Pwani', 520),
 ('Amanzi', 'Abidjan', 400),
 ('Amanzi', 'Bello', 400),
 ('Hawassa', 'Rural', 3900),
 ('Hawassa', 'Serowe', 480),
 ('Hawassa', 'Amina', 420),
 ('Hawassa', 'Deka', 350),
 ('Hawassa', 'Djenne', 310),
 ('Hawassa', 'Zanzibar', 300),
 ('Hawassa', 'Yaounde', 270),
 ('Kilimani', 'Rural', 5440),
 ('Kilimani', 'Mrembo', 990),
 ('Kilimani', 'Harare', 850),
 ('Kilimani', 'Isiqalo', 770),
 ('Kilimani', 'Zuri', 750),
 ('Kilimani', 'Amara', 710),
 ('Sokoto', 'Rural', 5010),
 ('Sokoto', 'Ilanga', 780),
 ('Sokoto', 'Majengo', 620),
 ('Sokoto', 'Kofi', 510),
 ('Sokoto', 'Bahari', 470),
 ('Sokoto', 'Marang', 460),
 ('Sokoto', 'Cheche', 370)]

### Counting Records per Location Type

Having verified that our survey data is both comprehensive and consistent, we can now analyze the **types of water sources** captured across Maji Ndogo.

By counting the number of records per `location_type`, we gain valuable insights into:
- The dominant types of water sources (e.g., wells, taps, boreholes, rivers).
- How evenly these sources are distributed across the dataset.
- The diversity and accessibility of water infrastructure in the region.

This will help decision-makers understand which water source types are most common — and which might need more attention or investment.


In [11]:
%%sql

SELECT
    location_type,
    COUNT(*) AS num_records
FROM location
GROUP BY location_type
ORDER BY num_records DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
2 rows affected.


[('Rural', 23740), ('Urban', 15910)]

### Calculating Rural vs. Urban Water Source Percentages

While counts tell us how many water sources exist, percentages give us a clearer sense of proportion — which helps in communicating impact and equity.

To calculate the proportion of rural sources in our dataset, we can use SQL as a simple calculator.  
In our case:

```sql
SELECT 23740 / (15910 + 23740) * 100;


###  Insights from the Location Data

From our analysis of the `location` table, several key insights emerge that strengthen our understanding of the Maji Ndogo water landscape:



####  Comprehensive Coverage
Every province and town has been well-documented, meaning our dataset provides an accurate reflection of water access conditions across Maji Ndogo.



####  Rural Majority
Around **60% of the water sources** are found in **rural areas**, highlighting that the majority of Maji Ndogo’s population depends on **non-urban water infrastructure**.



####  Data Integrity & Representativeness
The balanced distribution across provinces and towns increases confidence in the dataset’s **reliability**, making it a solid foundation for **evidence-based decision-making**.



These findings demonstrate that our surveyors achieved **broad and accurate coverage**, giving us high confidence that our data truly represents the conditions on the ground.


##  Diving into the Water Sources

The `water_source` table contains detailed information about the various types of water sources across Maji Ndogo — including wells, taps, rivers, and more.  
This dataset helps us understand **how people access water**, **what types of sources exist**, and **how widely each source is used**.

Before running any calculations, let’s explore the structure of the table to understand its key columns and data types.



###  Key Questions to Explore

1. **How many people did we survey in total?**  
   This helps us understand the dataset’s coverage and the population represented.

2. **How many wells, taps, and rivers are there?**  
   This shows us the distribution of different water source types.

3. **How many people share each type of water source on average?**  
   This will reveal how heavily shared or strained some water sources might be.

4. **How many people are getting water from each type of source?**  
   This provides insight into population access by source type.



Let’s start by addressing the **second question** — counting how many of each water source type exists.


In [16]:
%%sql

SELECT
   *
FROM water_source
limit 5;

 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486),
 ('AkHa00003224', 'well', 364),
 ('AkHa00004224', 'tap_in_home_broken', 942)]

###  Counting Water Source Types

Now that we’ve examined the structure of the `water_source` table, we can start answering our second question:

> **How many wells, taps, and rivers are there?**

By grouping the data by `type_of_water_source`, we can see how many of each type of source exists across Maji Ndogo.  
This helps us understand the **availability and diversity** of water infrastructure in different communities.


In [17]:
%%sql

SELECT
    type_of_water_source,
    COUNT(*) AS num_sources
FROM water_source
GROUP BY type_of_water_source
ORDER BY num_sources DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[('well', 17383),
 ('tap_in_home', 7265),
 ('tap_in_home_broken', 5856),
 ('shared_tap', 5767),
 ('river', 3379)]

###  Average Number of People Served per Water Source

Now that we’ve identified the types and counts of water sources in Maji Ndogo, the next step is to understand **how heavily each source is used**.

> **Question:** What is the average number of people that are served by each water source type?

This insight helps us understand how demand is distributed — for instance, whether **taps**, **wells**, or **rivers** serve more people on average.

If one source type consistently serves a higher number of people, it may be a critical point of pressure in the water system and should be prioritized during maintenance and upgrades.

We'll calculate the **average number of people served per water source type**, rounding the results for easier interpretation.


In [18]:
%%sql

SELECT
    type_of_water_source,
    ROUND(AVG(number_of_people_served), 0) AS avg_people_served
FROM
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    avg_people_served DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[('shared_tap', Decimal('2071')),
 ('river', Decimal('699')),
 ('tap_in_home_broken', Decimal('649')),
 ('tap_in_home', Decimal('644')),
 ('well', Decimal('279'))]

###  Total Number of People Served per Water Source Type

Our previous query gave us the **average number of people** served per source type, but averages alone can be misleading.  

For instance:
- The data showed that *tap_in_home* sources serve **around 644 people on average** — which clearly doesn’t make sense.  
  In reality, each household has its **own tap**, and with an **average household size of 6**, that means:
  
  > 1 recorded `tap_in_home` ≈ 100 actual taps (644 ÷ 6 ≈ 100)

This illustrates why it’s essential to **question the data** — not just calculate.  
Analysts who interpret context add real value to decision-making.

Now, instead of focusing on the average, we’ll calculate the **total number of people served** by each source type.  
This helps us identify which sources have the **greatest population impact** and should be prioritized for repairs or expansion.


In [19]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served
FROM
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    total_people_served DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272')),
 ('well', Decimal('4841724')),
 ('tap_in_home', Decimal('4678880')),
 ('tap_in_home_broken', Decimal('3799720')),
 ('river', Decimal('2362544'))]

###  Insights

- **Public taps** serve the largest number of people overall — confirming the long queues we observed earlier.  
- **Tap_in_home** sources appear inflated because of data grouping — remember, each record actually represents many households.  
- **Wells and boreholes** serve fewer people individually, but remain vital in remote areas with limited infrastructure.  
- When planning repairs or upgrades, we should **prioritize shared sources** (public taps, community wells, etc.), since improvements there will benefit the largest number of citizens.


###  Calculating the Percentage of Citizens Served by Each Water Source Type

The total population surveyed in Maji Ndogo is approximately **27 million citizens**.

While the total counts of people served by each water source show us raw scale, percentages make it **easier to understand relative impact** — i.e., what share of the population depends on each water source type.

We’ll calculate:

\[
\text{Percentage Served} = \left( \frac{\text{Total People Served per Source}}{27,000,000} \right) \times 100
\]

This will show which water source types account for the majority of water access in Maji Ndogo.


In [20]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served,
    ROUND(SUM(number_of_people_served) / 27000000 * 100, 2) AS percentage_of_population
FROM
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    total_people_served DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272'), Decimal('44.24')),
 ('well', Decimal('4841724'), Decimal('17.93')),
 ('tap_in_home', Decimal('4678880'), Decimal('17.33')),
 ('tap_in_home_broken', Decimal('3799720'), Decimal('14.07')),
 ('river', Decimal('2362544'), Decimal('8.75'))]

###  Insights from Percentage Breakdown

- The **dominant water source type** accounts for a significant share of the population — showing strong dependence on a few key systems.  
- **Public taps** likely serve the largest proportion, indicating a high reliance on shared water infrastructure.  
- **Tap_in_home** sources serve a smaller proportion (once corrected for data aggregation), while **wells** and **rivers** cover rural or remote populations.  
- This percentage-based view helps policymakers **prioritize infrastructure repairs and investments** for the most impactful improvements.


###  Simplifying Percentages for Better Communication

Having percentages with multiple decimal places doesn’t make the insights clearer — it actually makes them harder to read.

To improve readability, let’s **round the percentages to 0 decimal places** and **order the results** so that the largest water source types appear first. This helps decision-makers quickly grasp which sources serve the most people.


In [21]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS total_people_served,
    ROUND(SUM(number_of_people_served) / 27000000 * 100, 0) AS percentage_of_population
FROM
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    total_people_served DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272'), Decimal('44')),
 ('well', Decimal('4841724'), Decimal('18')),
 ('tap_in_home', Decimal('4678880'), Decimal('17')),
 ('tap_in_home_broken', Decimal('3799720'), Decimal('14')),
 ('river', Decimal('2362544'), Decimal('9'))]

###  Key Takeaways

- Rounding percentages to whole numbers gives a clearer picture of how water sources are distributed.  
- We can now easily identify which water sources account for the largest shares of access.  
- This simplified data presentation is ideal for **reports, dashboards, and stakeholder briefings**.


###  Ranking Water Source Types by Total People Served (excluding `tap_in_home`)

We want to prioritise fixes that will help the most people. To do that, we:

1. Compute the **total number of people served** by each `type_of_water_source`.  
2. **Exclude** `tap_in_home` from the ranking because homes already have the best source available and are not candidates for infrastructure prioritisation.  
3. Use a SQL **window function** (`RANK()`) on the aggregated totals to produce a ranked list (rank 1 = highest total people served).

This result will help decision-makers quickly identify which source types (e.g., shared taps, rivers, wells) impact the largest number of citizens and should therefore be prioritised for repairs or upgrades.


In [22]:
%%sql


-- Rank source types by total people served, excluding tap_in_home
SELECT
    type_of_water_source,
    total_people_served,
    RANK() OVER (ORDER BY total_people_served DESC) AS rank_by_people_served
FROM (
    SELECT
        type_of_water_source,
        SUM(number_of_people_served) AS total_people_served
    FROM water_source
    WHERE type_of_water_source <> 'tap_in_home'
    GROUP BY type_of_water_source
) AS totals
ORDER BY total_people_served DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
4 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home_broken', Decimal('3799720'), 3),
 ('river', Decimal('2362544'), 4)]

###  Prioritising Individual Water Sources for Repair

Now that we know **which types of sources** serve the most people, let’s identify **specific sources** that should be prioritised for repairs.

We’ll apply the same logic:  
> The most-used sources within each type should be fixed first.

#### Requirements:
1. Assign a **rank within each water source type** based on the total number of people served.  
2. Limit the results to **improvable sources only** (for example, exclude fully functional ones like `tap_in_home`).  
3. Use a **window function** (`RANK()`) with `PARTITION BY` to rank within each type.  
4. Order the final results so the most critical sources appear at the top.

This query helps planners make data-driven repair schedules by focusing resources where they’ll have the greatest impact.


In [23]:
%%sql


-- Rank individual water sources within each improvable type by number of people served
SELECT
    type_of_water_source,
    source_id,
    number_of_people_served,
    RANK() OVER (
        PARTITION BY type_of_water_source
        ORDER BY number_of_people_served DESC
    ) AS rank_within_type
FROM water_source
WHERE type_of_water_source IN (
    'shared_tap',
    'tap_in_home_broken',
    'well',
    'river'
)
ORDER BY
    type_of_water_source,
    rank_within_type
LIMIT 20;


 * mysql+pymysql://root:***@localhost/md_water_services
20 rows affected.


[('river', 'SoRu34798224', 998, 1),
 ('river', 'SoRu35837224', 998, 1),
 ('river', 'SoRu36238224', 998, 1),
 ('river', 'SoRu36791224', 998, 1),
 ('river', 'SoRu36880224', 998, 1),
 ('river', 'SoRu38142224', 998, 1),
 ('river', 'SoRu37756224', 998, 1),
 ('river', 'SoMa33775224', 998, 1),
 ('river', 'KiRu30353224', 998, 1),
 ('river', 'SoIl32972224', 998, 1),
 ('river', 'KiZu31236224', 998, 1),
 ('river', 'KiRu28591224', 998, 1),
 ('river', 'KiMr25030224', 998, 1),
 ('river', 'KiRu25603224', 996, 14),
 ('river', 'KiRu26027224', 996, 14),
 ('river', 'KiRu30532224', 996, 14),
 ('river', 'SoRu36229224', 996, 14),
 ('river', 'AkRu08053224', 996, 14),
 ('river', 'AmAb09070224', 996, 14),
 ('river', 'SoKo33179224', 994, 20)]

###  Comparing RANK(), DENSE_RANK(), and ROW_NUMBER()

Now that we’re ranking water sources to prioritize repairs, it’s important to understand the **differences between SQL ranking functions** and when to use each:

| Function | Description | How It Handles Ties | Example Use Case |
|-----------|--------------|--------------------|------------------|
| **RANK()** | Assigns ranks with gaps after ties. | Equal values share the same rank; the next rank is skipped. | When you want to show true rank order but preserve ties (e.g., two sources equally prioritized). |
| **DENSE_RANK()** | Similar to `RANK()` but **without gaps** in rank numbers. | Equal values share the same rank, but no gaps. | When you want continuous ranks that are easy to interpret by teams. |
| **ROW_NUMBER()** | Assigns a **unique rank** to every row. | No ties — each record gets its own number. | When exact order matters (e.g., task lists where engineers fix one source after another). |

####  Practical Takeaway
- **RANK()** is great for analysis and reporting.  
- **DENSE_RANK()** makes the priority list easier to understand by avoiding skipped numbers.  
- **ROW_NUMBER()** gives a unique ID for operations or task tracking — ideal for engineers fixing specific sources.

Try running all three ranking functions to see how they behave with your dataset — it’s a great way to build intuition about SQL window functions.


###  Analysing Queues — How Long Did the Survey Take?

Let’s begin our **queue analysis** by determining how long the entire survey lasted.

The `visits` table contains a column called `time_of_record`, which stores timestamps of each field visit.

To find the **survey duration**, we’ll:
1. Identify the **earliest** (`MIN`) and **latest** (`MAX`) recorded dates.
2. Use the `DATEDIFF()` function to calculate the **number of days** between these two dates.

This will tell us how long our field teams were actively collecting data.


In [24]:
%%sql


-- Calculate how long the survey took (in days)
SELECT
    MIN(time_of_record) AS survey_start_date,
    MAX(time_of_record) AS survey_end_date,
    DATEDIFF(MAX(time_of_record), MIN(time_of_record)) AS survey_duration_days
FROM visits;


 * mysql+pymysql://root:***@localhost/md_water_services
1 rows affected.


[(datetime.datetime(2021, 1, 1, 9, 10), datetime.datetime(2023, 7, 14, 13, 53), 924)]

###  Reflecting on the Human Side of Data

Just imagine all the visits, and meeting all those people on the ground — **over two years** of data collection!  
It’s easy to view these figures as just numbers, but each record represents a real person who may have:

- Walked **10 kilometres** to reach a water source.  
- Waited **4–5 hours** in line for water.  
- Then walked **all the way back home**, often under harsh conditions.

Many of those queuing were **children**, missing out on school or playtime.  
It’s a sobering reminder that behind every data point is a story of resilience — and a reason for us to make things better.

> “Mambo yatakuwa sawa,” as my mother used to say — *things will be okay.*  
> Our responsibility is to make sure the data we collect helps prevent this hardship in the future.


###  Question 2 — Average Queue Time for Water Collection

Next, let’s find out how long people have to queue on average in Maji Ndogo.  
However, not all water sources have queues — for instance, **taps in homes** have a `time_in_queue` value of `0`.

If we include these zeroes in our average calculation, they’ll **skew** the results and make it seem like people wait less than they actually do.

To solve this, we can use the **`NULLIF()`** function, which converts zeros into `NULL` values (and SQL automatically ignores `NULL`s in aggregations like `AVG()`).

This will give us a **realistic picture** of queue times for people who actually had to wait for water.


In [25]:
%%sql


-- Calculate the average queue time, excluding 0-minute queues
SELECT
    ROUND(AVG(NULLIF(time_in_queue, 0)), 2) AS avg_queue_time_minutes
FROM
    visits;


 * mysql+pymysql://root:***@localhost/md_water_services
1 rows affected.


[(Decimal('123.26'),)]

###  Insight

- The result represents the **average queue time (in minutes)** for citizens who actually experienced a queue.  
- This helps us understand **how severe waiting times are** across Maji Ndogo.  
- We can later combine this with location data to identify **which provinces or towns** have the longest queues — guiding targeted interventions.


###  Question 3 — Average Queue Time by Day of the Week

Now that we know the **average queue time** is about **123 minutes**, let’s explore how this varies depending on the **day of the week**.

People might prefer collecting water on specific days — for example:
- **Weekends**, when they’re not working or attending school.
- **Mondays**, when households refill after the weekend.

To analyze this, we’ll use the `DAYNAME()` function, which extracts the **day of the week** (e.g., Monday, Tuesday) from a timestamp column.

We’ll calculate the **average queue time per day**, excluding 0-minute queues (for taps in homes).


In [26]:
%%sql


-- Calculate average queue time for each day of the week
SELECT
    DAYNAME(time_of_record) AS day_of_week,
    ROUND(AVG(NULLIF(time_in_queue, 0)), 2) AS avg_queue_time_minutes
FROM
    visits
GROUP BY
    day_of_week
ORDER BY
    avg_queue_time_minutes DESC;


 * mysql+pymysql://root:***@localhost/md_water_services
7 rows affected.


[('Saturday', Decimal('246.29')),
 ('Monday', Decimal('136.63')),
 ('Friday', Decimal('119.66')),
 ('Tuesday', Decimal('107.94')),
 ('Thursday', Decimal('105.38')),
 ('Wednesday', Decimal('96.62')),
 ('Sunday', Decimal('81.53'))]

###  Interpretation

- The `DAYNAME()` function extracts the weekday name from each record’s timestamp.  
- We grouped by these names to find the **average waiting time per day**.

By ordering the results in descending order, we can easily see **which days have the longest queues**.

 **Insight Example (if your result matches expectations):**
- Queue times tend to **peak on weekends (Saturday or Sunday)** when most people are off work or school.  
- **Midweek days (Tuesday–Thursday)** may have shorter queues, indicating more balanced water collection patterns.

This tells us when water demand is **highest**, helping planners **schedule maintenance or allocate resources** more effectively.


###  Question 4 — Queue Time by Hour of the Day

Next, let’s analyze **what time of day** people typically collect water.

We can extract the **hour** from the `time_of_record` column using the `HOUR()` function.  
This gives us an integer value (0–23), representing the hour of the day when the record was taken.

By calculating the **average queue time per hour**, we can visualize *daily water-fetching patterns*.

For instance:
- Early morning (5–8 AM) might be busy before school/work.
- Late afternoon (4–7 PM) could also be busy as people return home.


In [27]:
%%sql


-- Calculate average queue time by hour of the day
SELECT
    HOUR(time_of_record) AS hour_of_day,
    ROUND(AVG(NULLIF(time_in_queue, 0)), 2) AS avg_queue_time_minutes
FROM
    visits
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day;


 * mysql+pymysql://root:***@localhost/md_water_services
14 rows affected.


[(6, Decimal('148.93')),
 (7, Decimal('149.13')),
 (8, Decimal('148.89')),
 (9, Decimal('117.76')),
 (10, Decimal('113.89')),
 (11, Decimal('110.69')),
 (12, Decimal('111.52')),
 (13, Decimal('115.05')),
 (14, Decimal('114.24')),
 (15, Decimal('114.21')),
 (16, Decimal('114.08')),
 (17, Decimal('148.79')),
 (18, Decimal('146.82')),
 (19, Decimal('167.70'))]

###  Interpretation

- The `HOUR()` function extracts the hour from each record’s timestamp.
- Grouping by this hour and averaging queue times shows **when queues are longest** during the day.
- Ordering by `hour_of_day` lets us see the pattern chronologically.

 **Example Insight (if your results follow a typical pattern):**
- Queues are **longest between 6–9 AM**, when families fetch water before work or school.
- Queues drop mid-day, then **rise again around 5–7 PM**, as people return home.
- Very early or late hours show little to no activity.

This information helps water authorities **plan collection schedules**, or **adjust staffing and maintenance** to match peak demand.


###  Question 5 — Average Queue Time per Hour, per Day of Week

Let’s create a pivot-style table showing the **average queue time (minutes)** for each **hour of the day**, broken down by **day of the week**.

In SQL, there’s no built-in `PIVOT()` function in most databases, so we can simulate it using:
- `CASE()` → to filter values for each day into their own column.
- `AVG()` → to compute the average queue time for that day/hour.
- `ROUND()` → to make the output easy to read.
- `GROUP BY` → to aggregate by `hour_of_day`.

We’ll also format the hour nicely using `TIME_FORMAT()`.


In [28]:
%%sql


SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,

    -- Sunday
    ROUND(AVG(
        CASE WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue ELSE NULL END
    ), 0) AS Sunday,

    -- Monday
    ROUND(AVG(
        CASE WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue ELSE NULL END
    ), 0) AS Monday,

    -- Tuesday
    ROUND(AVG(
        CASE WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue ELSE NULL END
    ), 0) AS Tuesday,

    -- Wednesday
    ROUND(AVG(
        CASE WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue ELSE NULL END
    ), 0) AS Wednesday,

    -- Thursday
    ROUND(AVG(
        CASE WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue ELSE NULL END
    ), 0) AS Thursday,

    -- Friday
    ROUND(AVG(
        CASE WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue ELSE NULL END
    ), 0) AS Friday,

    -- Saturday
    ROUND(AVG(
        CASE WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue ELSE NULL END
    ), 0) AS Saturday

FROM
    visits
WHERE
    time_in_queue != 0
GROUP BY
    hour_of_day
ORDER BY
    hour_of_day;


 * mysql+pymysql://root:***@localhost/md_water_services
14 rows affected.


[('06:00', Decimal('79'), Decimal('190'), Decimal('134'), Decimal('112'), Decimal('134'), Decimal('153'), Decimal('247')),
 ('07:00', Decimal('82'), Decimal('186'), Decimal('128'), Decimal('111'), Decimal('139'), Decimal('156'), Decimal('247')),
 ('08:00', Decimal('86'), Decimal('183'), Decimal('130'), Decimal('119'), Decimal('129'), Decimal('153'), Decimal('247')),
 ('09:00', Decimal('84'), Decimal('127'), Decimal('105'), Decimal('94'), Decimal('99'), Decimal('107'), Decimal('252')),
 ('10:00', Decimal('83'), Decimal('119'), Decimal('99'), Decimal('89'), Decimal('95'), Decimal('112'), Decimal('259')),
 ('11:00', Decimal('78'), Decimal('115'), Decimal('102'), Decimal('86'), Decimal('99'), Decimal('104'), Decimal('236')),
 ('12:00', Decimal('78'), Decimal('115'), Decimal('97'), Decimal('88'), Decimal('96'), Decimal('109'), Decimal('239')),
 ('13:00', Decimal('81'), Decimal('122'), Decimal('97'), Decimal('98'), Decimal('101'), Decimal('115'), Decimal('242')),
 ('14:00', Decimal('83'), Decimal('127'), Decimal('104'), Decimal('92'), Decimal('96'), Decimal('110'), Decimal('244')),
 ('15:00', Decimal('83'), Decimal('126'), Decimal('104'), Decimal('88'), Decimal('92'), Decimal('110'), Decimal('248')),
 ('16:00', Decimal('83'), Decimal('127'), Decimal('99'), Decimal('90'), Decimal('99'), Decimal('109'), Decimal('251')),
 ('17:00', Decimal('79'), Decimal('181'), Decimal('135'), Decimal('121'), Decimal('129'), Decimal('151'), Decimal('251')),
 ('18:00', Decimal('80'), Decimal('174'), Decimal('122'), Decimal('113'), Decimal('132'), Decimal('158'), Decimal('240')),
 ('19:00', Decimal('127'), Decimal('159'), Decimal('145'), Decimal('176'), Decimal('137'), Decimal('103'), Decimal('282'))]

###  Interpretation

Now we have a table where:
- Each **row** = hour of the day (e.g. 06:00, 07:00, 08:00)
- Each **column** = average queue time on that **day of the week**

From this, we can spot:
- Which **hours** have the longest queues.
- Which **days** are most congested.
- How **weekend patterns** differ from **weekdays**.

 For example:
- Early mornings (6–8 AM) and evenings (5–7 PM) may show consistently high queue times.
- Sundays or Fridays might have lighter queues as people rest or pray.
- Midday hours (11 AM–2 PM) are often quieter.

This kind of breakdown helps **plan maintenance schedules** or **staff rotations** around real community behavior.


# Maji Ndogo — Water Accessibility & Infrastructure Summary (Executive Brief)

**To:** Pres. Aziza Naledi  
**From:** Data Team — Maji Ndogo Survey  
**Date:** 2025-10-29



## Key Insights

1. **Most sources are rural.**  
   The survey shows water sources are widely distributed across rural communities — our interventions must prioritise rural logistics.

2. **Shared taps dominate (43%).**  
   Approximately **43%** of people rely on shared community taps. These are high-impact targets for repair and temporary relief.

3. **Home taps exist but many systems are non-functional.**  
   About **31%** of people have `tap_in_home` or `tap_in_home_broken` recorded. Of those, **~45%** are connected to non-functional infrastructure (pipes, pumps, reservoirs).

4. **Wells remain important but many are unclean.**  
   **18%** of people use wells; however, only **~28%** of sampled wells were classified as clean.

5. **Long queue times — serious daily burden.**  
   Average queue time for people who queue is **~123 minutes** (about 2 hours). This represents a substantial daily time cost, often borne by children and caregivers.

6. **Queue timing patterns:**  
   - Longest queues: **Saturdays**, and **mornings/evenings** on many weekdays.  
   - Shortest queues: **Sundays** (cultural practices).  
   These patterns are actionable for both short-term relief (tankers) and scheduling maintenance.



## Prioritisation Framework (Data-driven)

We recommend prioritising fixes that benefit the **most people** first:

1. **Top priority — Shared taps**  
   - Largest total population affected.  
   - High immediate impact when repaired or temporarily supplemented.

2. **Second — Contaminated wells**  
   - Many people rely on wells; cleaning/filtration yields large health benefits.

3. **Third — Broken communal infrastructure (reservoirs, pumps, pipes)**  
   - Fixing a single facility can restore service to many connected taps.

4. **Lower priority (for now) — Installing home taps at scale**  
   - High-impact but very resource-intensive. Focus on shared-source relief first where queue times are high.



## Practical Actions (Short & Medium Term)

**Short term (immediate relief)**  
- Deploy water tankers to busiest shared taps on peak days/hours (use pivot table to schedule).  
- Provide portable filtration or bottled water where contamination is severe.

**Medium term (targeted repairs & upgrades)**  
- Prioritise repairs for top-ranked shared taps and contaminated wells (use ranked lists by `number_of_people_served`).  
- Repair distribution infrastructure (reservoirs, pumps, pipes) where multiple taps are affected.

**Long term (sustainable access)**  
- Plan for phased household tap installations focused on areas with persistently high queue times and strong logistics support.  
- Investigate root causes of well contamination and implement watershed/ sanitation interventions.



## Targets & Metrics

- **Goal:** Reduce average queue time to **≤ 30 minutes** (UN standard).  
- **Priority metric:** people helped per intervention (maximize).  
- **Operational KPIs:** # of taps/wells repaired, population served per intervention, avg queue time (pre/post), # of contaminated wells cleaned.



## Next Steps (Immediate)

1. Finalise the **top-ranked list of improvable sources** (shared taps, broken communal infrastructure, contaminated wells).  
2. Request **cost estimates** for:
   - Repairing shared taps / distribution nodes
   - Drilling wells / installing household taps
   - Installing filtration (UV/RO) for contaminated wells  
3. Coordinate **logistics assessment** for rural deployments (road access, materials, crews).  
4. Begin targeted **pilot repairs** at highest-ranked shared taps and measure queue time improvement.



## Closing

This dataset gives us a strong, representative foundation to prioritise interventions. By focusing on shared taps and contaminated wells (where we will get the most immediate impact), and scheduling relief around peak queue days/hours, we can both reduce suffering quickly and plan sustainable upgrades.

I’ll prepare the top-ranked source list and a template email for requesting cost estimates next. Please let me know if you’d like me to draft the message to potential contractors or to the Ministry of Public Works.
